python/ray/serve/tests/test_proxy.py

```diff
@@ -1,11 +1,681 @@
+            )
+        mocked_proxy_request.assert_called_once()
+
+    @pytest.mark.asyncio
+    # @pytest.mark.skip(reason="WIP: TODO (genesu)")
+    @patch("ray.serve._private.http_proxy.receive_http_body")
+    async def test_send_request_to_replica_unary(self, mock_receive_http_body):
+        """Test HTTPProxy send_request_to_replica_unary returns the correct response."""
+
+        http_body_bytes = b""
+        mock_receive_http_body.return_value = http_body_bytes
+
+        async def async_obj():
+            return "foo"
+
+        MagicMock.__await__ = lambda x: async_obj().__await__()
+        mocked_obj_ref = MagicMock()
+
+        async def async_assignment_task():
+            return mocked_obj_ref
+
+        async def receive_message():
+            return {"type": "http.disconnect"}
+
+        MagicMock.__await__ = lambda x: async_assignment_task().__await__()
+        mock_assignment_task = MagicMock()
+        handle = MagicMock()
+        handle.remote = mock_assignment_task
+        proxy_request = ASGIProxyRequest(
+            scope={},
+            receive=receive_message,
+            send=AsyncMock(),
+        )
+        http_proxy = self.create_http_proxy()
+        returned_response = await http_proxy.send_request_to_replica_unary(
+            handle=handle,
+            proxy_request=proxy_request,
+        )
+
+        assert isinstance(returned_response, ProxyResponse)
+        assert returned_response.status_code == http_proxy.success_status_code
+
+    @pytest.mark.asyncio
+    async def test_proxy_asgi_receive(self):
+        """Test HTTPProxy proxy_asgi_receive receives messages."""
+        http_proxy = self.create_http_proxy()
+        receive = AsyncMock()
+        receive.side_effect = [
+            {"type": "http.request"},
+            {"type": "http.request"},
+            {"type": "http.disconnect"},
+        ]
+        queue = AsyncMock()
+        await http_proxy.proxy_asgi_receive(
+            receive=receive,
+            queue=queue,
+        )
+
+        queue.close.assert_called_once()
+
+    @pytest.mark.asyncio
+    async def test_consume_and_send_asgi_message_generator(self):
+        """Test HTTPProxy _consume_and_send_asgi_message_generator consumes the correct
+        generator, sends the message, and returns the correct status code.
+        """
+        http_proxy = self.create_http_proxy()
+        status_code = "status_code"
+        asgi_message = {"type": "http.response.start", "status": status_code}
+        asgi_messages = [asgi_message]
+        response_bytes = pickle.dumps(asgi_messages)
+
+        async def async_magic():
+            return response_bytes
+
+        MagicMock.__await__ = lambda x: async_magic().__await__()
+        obj_ref = MagicMock()
+        obj_ref.is_nil = MagicMock(return_value=False)
+        obj_ref_generator = AsyncMock()
+        obj_ref_generator._next_async.side_effect = [
+            obj_ref,
+            StopAsyncIteration(),
+        ]
```

Comment:
let's not mock `obj_ref` directly, if we need to do this we should wrap it in a higher-level interface and then write a fake for that interface